signature/internal/base/envelope.go

```diff
@@ -0,0 +1,221 @@
+package base
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/signature"
+	nx509 "github.com/notaryproject/notation-core-go/x509"
+)
+
+// Envelope represents a general envelope wrapping a raw signature and envelope
+// in specific format.
+// Envelope manipulates the common validation shared by internal envelopes.
+type Envelope struct {
+	signature.Envelope        // internal envelope in a specific format(e.g. Cose, JWS)
+	Raw                []byte // raw signature
+}
+
+// Sign generates signature in terms of given SignRequest.
+//
+// Reference: https://github.com/notaryproject/notaryproject/blob/main/signing-and-verification-workflow.md#signing-steps
+func (e *Envelope) Sign(req *signature.SignRequest) ([]byte, error) {
+	// Canonicalize request.
+	req.SigningTime = req.SigningTime.Truncate(time.Second)
+	req.Expiry = req.Expiry.Truncate(time.Second)
+	err := validateSignRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	raw, err := e.Envelope.Sign(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// validate certificate chain
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+	signerInfo := content.SignerInfo
+
+	if err := validateCertificateChain(
+		signerInfo.CertificateChain,
+		signerInfo.SignedAttributes.SigningTime,
+		signerInfo.SignatureAlgorithm,
+	); err != nil {
+		return nil, err
+	}
+
+	e.Raw = raw
+	return e.Raw, nil
+}
+
+// Verify performs integrity and other signature specification related
+// validations.
+// It returns envelope content containing the payload to be signed and
+// SignerInfo object containing the information about the signature.
+//
+// Reference: https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#steps
+func (e *Envelope) Verify() (*signature.EnvelopeContent, error) {
+	// validation before the core verify process.
+	if len(e.Raw) == 0 {
+		return nil, &signature.SignatureNotFoundError{}
+	}
+
+	// core verify process.
+	content, err := e.Envelope.Verify()
+	if err != nil {
+		return nil, err
+	}
+
+	// validation after the core verify process.
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// Content returns the validated signature information and payload.
+func (e *Envelope) Content() (*signature.EnvelopeContent, error) {
+	if len(e.Raw) == 0 {
+		return nil, &signature.SignatureNotFoundError{}
+	}
+
+	content, err := e.Envelope.Content()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = validateEnvelopeContent(content); err != nil {
+		return nil, err
+	}
+
+	return content, nil
+}
+
+// validateSignRequest performs basic set of validations on SignRequest struct.
+func validateSignRequest(req *signature.SignRequest) error {
+	if err := validatePayload(&req.Payload); err != nil {
```

Comment:
This returns `InvalidSignatureError` which is not correct since signature is not yet generated.

I think we will need to verify this for all the validations done in `validateSignRequest`

Reply:
Thanks for catching it. I have updated to return `InvalidSignRequestError` in `validateSignRequest` instead. But I'll keep this validation stay here.